app/routes.py

```diff
@@ -1 +1,195 @@
-from flask import Blueprint
\ No newline at end of file
+import datetime
+import requests
+import os
+from app import db
+from app.models.task import Task
+from app.models.goal import Goal
+from flask import Blueprint, jsonify, make_response, abort, request
+
+goal_bp = Blueprint("goals", __name__, url_prefix="/goals")
+task_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+
+SLACKBOT_TOKEN = os.environ.get("SLACKBOT_TOKEN")
+
+
+def validate_model(cls, id):
+    try:
+        model_id = int(id)
+    except:
+        abort(make_response({"message":f"{cls.__name__} {id} invalid"}, 400))
+
+    model = cls.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message":f"{cls.__name__} {id} not found"}, 404))
+
+    return model
+
+#==============================
+#         GOAL ROUTES
+#==============================
+@goal_bp.route("/<goal_id>", methods=["GET"])
+def read_one_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+    return {"goal": goal.to_dict()}
+
+@goal_bp.route("", methods=["POST"])
+def create_goal():
+    request_body = request.get_json()
+    new_goal = Goal.instance_from_json(request_body)
+
+    db.session.add(new_goal)
+    db.session.commit()
+
+    return make_response(jsonify({"goal": new_goal.to_dict()}), 201)
+
+@goal_bp.route("", methods=["GET"])
+def read_all_goals():
+    title_query = request.args.get("title")
+    if title_query:
+        goals = Goal.query.filter_by(title=title_query)
+    else:
+        goals = Goal.query.all()
+
+    goals_response = [goal.to_dict() for goal in goals]
+    return jsonify(goals_response)
+
+@goal_bp.route("/<goal_id>", methods=["PUT"])
+def update_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+
+    request_body = request.get_json()
+
+    goal.update(request_body)
+
+    db.session.commit()
+
+    return {"goal": goal.to_dict()}
+
+@goal_bp.route("/<goal_id>", methods=["DELETE"])
+def delete_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+
+    db.session.delete(goal)
+    db.session.commit()
+
+    return make_response(jsonify({"details": f'Goal {goal.goal_id} "{goal.title}" successfully deleted'}))
+
+@goal_bp.route("/<goal_id>/tasks", methods=["POST"])
+def add_tasks_to_goal(goal_id):
+    goal = validate_model(Goal, goal_id)
+    request_body = request.get_json()
+    task_ids = request_body["task_ids"]
+
+    add_tasks_to_goal(task_ids, goal)
+
+    db.session.commit()
+
+    return make_response(jsonify({"id":goal.goal_id,"task_ids":task_ids}))
```

Comment:
Based on our tests, this works fine. But consider a goal whose tasks get added to repeatedly. If we we only ever return the tasks that have most recently been added, we might miss out on previous task ids that were already there.

Consider something like this instead:
```python
return make_response(jsonify({"id":goal.goal_id,"task_ids":[task.task_id for task in goal.many_tasks]}))
```